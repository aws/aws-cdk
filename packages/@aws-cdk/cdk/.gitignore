*.js
*.d.ts
tsconfig.json
tslint.json
*.js.map
dist
coverage
.nyc_output
.jsii

.LAST_BUILD
.nycrc
.LAST_PACKAGE
*.snk