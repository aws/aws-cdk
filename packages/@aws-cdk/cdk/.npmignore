
dist
.LAST_PACKAGE
*.tsbuildinfo
.LAST_BUILD
*.snk
*.ts
!*.d.ts
!*.js
coverage
.nyc_output
*.tgz

# Include .jsii
!.jsii
