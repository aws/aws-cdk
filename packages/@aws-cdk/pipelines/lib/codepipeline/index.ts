export * from './artifact-map';
export * from './codebuild-step';
export * from './codepipeline';
export * from './codepipeline-action-factory';
export * from './codepipeline-source';