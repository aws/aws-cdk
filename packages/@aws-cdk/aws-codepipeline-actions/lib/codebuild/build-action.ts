import * as codebuild from '@aws-cdk/aws-codebuild';
import * as codepipeline from '@aws-cdk/aws-codepipeline';
import * as iam from '@aws-cdk/aws-iam';
import * as cdk from '@aws-cdk/core';
import { Action } from '../action';

/**
 * The type of the CodeBuild action that determines its CodePipeline Category -
 * Build, or Test.
 * The default is Build.
 */
export enum CodeBuildActionType {
  /**
   * The action will have the Build Category.
   * This is the default.
   */
  BUILD,

  /**
   * The action will have the Test Category.
   */
  TEST
}

/**
 * Construction properties of the {@link CodeBuildAction CodeBuild build CodePipeline action}.
 */
export interface CodeBuildActionProps extends codepipeline.CommonAwsActionProps {
  /**
   * The source to use as input for this action.
   */
  readonly input: codepipeline.Artifact;

  /**
   * The list of additional input Artifacts for this action.
   */
  readonly extraInputs?: codepipeline.Artifact[];

  /**
   * The list of output Artifacts for this action.
   * **Note**: if you specify more than one output Artifact here,
   * you cannot use the primary 'artifacts' section of the buildspec;
   * you have to use the 'secondary-artifacts' section instead.
   * See https://docs.aws.amazon.com/codebuild/latest/userguide/sample-multi-in-out.html
   * for details.
   *
   * @default the action will not have any outputs
   */
  readonly outputs?: codepipeline.Artifact[];

  /**
   * The action's Project.
   */
  readonly project: codebuild.IProject;

  /**
   * The type of the action that determines its CodePipeline Category -
   * Build, or Test.
   *
   * @default CodeBuildActionType.BUILD
   */
  readonly type?: CodeBuildActionType;

  /**
   * The environment variables to pass to the CodeBuild project when this action executes.
   * If a variable with the same name was set both on the project level, and here,
   * this value will take precedence.
   *
   * @default - No additional environment variables are specified.
   */
  readonly environmentVariables?: { [name: string]: codebuild.BuildEnvironmentVariable };
}

/**
 * CodePipeline build action that uses AWS CodeBuild.
 */
export class CodeBuildAction extends Action {
  private readonly props: CodeBuildActionProps;

  constructor(props: CodeBuildActionProps) {
    super({
      ...props,
      category: props.type === CodeBuildActionType.TEST
        ? codepipeline.ActionCategory.TEST
        : codepipeline.ActionCategory.BUILD,
      provider: 'CodeBuild',
      artifactBounds: { minInputs: 1, maxInputs: 5, minOutputs: 0, maxOutputs: 5 },
      inputs: [props.input, ...props.extraInputs || []],
      resource: props.project,
    });

    this.props = props;
  }

  /**
   * Reference a CodePipeline variable defined by the CodeBuild project this action points to.
   * Variables in CodeBuild actions are defined using the 'exported-variables' subsection of the 'env'
   * section of the buildspec.
   *
   * @param variableName the name of the variable to reference.
   *   A variable by this name must be present in the 'exported-variables' section of the buildspec
   *
   * @see https://docs.aws.amazon.com/codebuild/latest/userguide/build-spec-ref.html#build-spec-ref-syntax
   */
  public variable(variableName: string): string {
    return this.variableExpression(variableName);
  }

  protected bound(scope: cdk.Construct, _stage: codepipeline.IStage, options: codepipeline.ActionBindOptions):
  codepipeline.ActionConfig {
    // check for a cross-account action if there are any outputs
    if ((this.actionProperties.outputs || []).length > 0) {
      const pipelineStack = cdk.Stack.of(scope);
      const projectStack = cdk.Stack.of(this.props.project);
      if (pipelineStack.account !== projectStack.account) {
        throw new Error('A cross-account CodeBuild action cannot have outputs. ' +
          'This is a known CodeBuild limitation. ' +
          'See https://github.com/aws/aws-cdk/issues/4169 for details');
      }
    }

    // grant the Pipeline role the required permissions to this Project
    options.role.addToPolicy(new iam.PolicyStatement({
      resources: [this.props.project.projectArn],
      actions: [
        'codebuild:BatchGetBuilds',
        'codebuild:StartBuild',
        'codebuild:StopBuild',
      ],
    }));

    // allow the Project access to the Pipeline's artifact Bucket
    // but only if the project is not imported
    // (ie., has a role) - otherwise, the IAM library throws an error
    if (this.props.project.role) {
      if ((this.actionProperties.outputs || []).length > 0) {
        options.bucket.grantReadWrite(this.props.project);
      } else {
        options.bucket.grantRead(this.props.project);
      }
    }

    if (this.props.project instanceof codebuild.Project) {
      this.props.project.bindToCodePipeline(scope, {
        artifactBucket: options.bucket,
      });
    }

    const configuration: any = {
      ProjectName: this.props.project.projectName,
      EnvironmentVariables: this.props.environmentVariables &&
        cdk.Stack.of(scope).toJsonString(codebuild.Project.serializeEnvVariables(this.props.environmentVariables)),
    };
    if ((this.actionProperties.inputs || []).length > 1) {
      // lazy, because the Artifact name might be generated lazily
      configuration.PrimarySource = cdk.Lazy.stringValue({ produce: () => this.props.input.artifactName });
    }
    return {
      configuration,
    };
  }
}
