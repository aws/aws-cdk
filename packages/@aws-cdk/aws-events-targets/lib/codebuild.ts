import * as codebuild from '@aws-cdk/aws-codebuild';
import * as events from '@aws-cdk/aws-events';
import * as iam from '@aws-cdk/aws-iam';
import { addToDeadLetterQueueResourcePolicy, bindBaseTargetConfig, singletonEventRole, TargetBaseProps } from './util';

/**
 * Customize the CodeBuild Event Target
 */
export interface CodeBuildProjectProps extends TargetBaseProps {

  /**
   * The role to assume before invoking the target
   * (i.e., the codebuild) when the given rule is triggered.
   *
   * @default - a new role will be created
   */
  readonly eventRole?: iam.IRole;

  /**
   * The event to send to CodeBuild
   *
   * This will be the payload for the StartBuild API.
   *
   * @default - the entire EventBridge event
   */
  readonly event?: events.RuleTargetInput;
}

/**
 * Start a CodeBuild build when an Amazon EventBridge rule is triggered.
 */
export class CodeBuildProject implements events.IRuleTarget {
  constructor(
    private readonly project: codebuild.IProject,
    private readonly props: CodeBuildProjectProps = {},
  ) {}

  /**
   * Allows using build projects as event rule targets.
   */
  public bind(_rule: events.IRule, _id?: string): events.RuleTargetConfig {

    if (this.props.deadLetterQueue) {
      addToDeadLetterQueueResourcePolicy(_rule, this.props.deadLetterQueue);
    }

    return {
      ...bindBaseTargetConfig(this.props),
      arn: this.project.projectArn,
      role: this.props.eventRole || singletonEventRole(this.project, [
        new iam.PolicyStatement({
          actions: ['codebuild:StartBuild'],
          resources: [this.project.projectArn],
        }),
      ]),
      input: this.props.event,
      targetResource: this.project,
    };
  }
}
