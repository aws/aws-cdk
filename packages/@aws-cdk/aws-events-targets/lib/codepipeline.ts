import * as codepipeline from '@aws-cdk/aws-codepipeline';
import * as events from '@aws-cdk/aws-events';
import * as iam from '@aws-cdk/aws-iam';
import { singletonEventRole } from './util';

/**
 * Customization options when creating a {@link CodePipeline} event target.
 */
export interface CodePipelineTargetOptions {
  /**
   * The role to assume before invoking the target
   * (i.e., the pipeline) when the given rule is triggered.
   *
   * @default - a new role will be created
   */
  readonly eventRole?: iam.IRole;
}

/**
 * Allows the pipeline to be used as an EventBridge rule target.
 */
export class CodePipeline implements events.IRuleTarget {
  constructor(
    private readonly pipeline: codepipeline.IPipeline,
    private readonly options: CodePipelineTargetOptions = {}) {
  }

  public bind(_rule: events.IRule, _id?: string): events.RuleTargetConfig {
    return {
      id: '',
      arn: this.pipeline.pipelineArn,
      role: this.options.eventRole || singletonEventRole(this.pipeline, [new iam.PolicyStatement({
        resources: [this.pipeline.pipelineArn],
        actions: ['codepipeline:StartPipelineExecution'],
      })]),
      targetResource: this.pipeline,
    };
  }
}
