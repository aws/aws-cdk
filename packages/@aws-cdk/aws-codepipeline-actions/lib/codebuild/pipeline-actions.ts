import codebuild = require('@aws-cdk/aws-codebuild');
import codepipeline = require('@aws-cdk/aws-codepipeline');
import iam = require('@aws-cdk/aws-iam');

/**
 * Common construction properties of all CodeBuild Pipeline Actions.
 */
export interface CommonCodeBuildActionProps extends codepipeline.CommonActionProps {
  /**
   * The source to use as input for this Action.
   */
  readonly inputArtifact: codepipeline.Artifact;

  /**
   * The list of additional input Artifacts for this Action.
   */
  readonly additionalInputArtifacts?: codepipeline.Artifact[];

  /**
   * The list of names for additional output Artifacts for this Action.
   * The resulting output artifacts can be accessed with the `additionalOutputArtifacts`
   * method of the Action.
   */
  readonly additionalOutputArtifactNames?: string[];

  /**
   * The Action's Project.
   */
  readonly project: codebuild.IProject;
}

/**
 * Construction properties of the {@link CodeBuildBuildAction CodeBuild build CodePipeline Action}.
 */
export interface CodeBuildBuildActionProps extends CommonCodeBuildActionProps {
  /**
   * The name of the build's output artifact.
   *
   * @default an auto-generated name will be used
   */
  readonly outputArtifactName?: string;
}

/**
 * CodePipeline build Action that uses AWS CodeBuild.
 */
export class CodeBuildBuildAction extends codepipeline.BuildAction {
  private readonly props: CodeBuildBuildActionProps;

  constructor(props: CodeBuildBuildActionProps) {
    super({
      ...props,
      provider: 'CodeBuild',
      artifactBounds: { minInputs: 1, maxInputs: 5, minOutputs: 0, maxOutputs: 5 },
      outputArtifactName: props.outputArtifactName || `Artifact_${props.actionName}_${props.project.node.uniqueId}`,
      configuration: {
        ProjectName: props.project.projectName,
      },
    });

    this.props = props;

    handleAdditionalInputOutputArtifacts(props, this,
      // pass functions to get around protected members
      this.actionInputArtifacts,
      (artifact) => this.addInputArtifact(artifact),
      (artifactName) => this.addOutputArtifact(artifactName));
  }

  /**
   * Returns the additional output artifacts defined for this Action.
   * Their names will be taken from the {@link CommonCodeBuildActionProps#additionalOutputArtifactNames}
   * property.
   *
   * @returns all additional output artifacts defined for this Action
   * @see #additionalOutputArtifact
   */
  public additionalOutputArtifacts(): codepipeline.Artifact[] {
    return this.actionOutputArtifacts.slice(1);
  }

  /**
   * Returns the additional output artifact with the given name,
   * or throws an exception if an artifact with that name was not found
   * in the additonal output artifacts.
   * The names are defined by the {@link CommonCodeBuildActionProps#additionalOutputArtifactNames}
   * property.
   *
   * @param name the name of the artifact to find
   * @returns the artifact with the given name
   * @see #additionalOutputArtifacts
   */
  public additionalOutputArtifact(name: string): codepipeline.Artifact {
    return findOutputArtifact(this.additionalOutputArtifacts(), name);
  }

  protected bind(info: codepipeline.ActionBind): void {
    setCodeBuildNeededPermissions(info, this.props.project, true);
  }
}

/**
 * Construction properties of the {@link CodeBuildTestAction CodeBuild test CodePipeline Action}.
 */
export interface CodeBuildTestActionProps extends CommonCodeBuildActionProps {
  /**
   * The optional name of the primary output artifact.
   * If you provide a value here,
   * then the `outputArtifact` property of your Action will be non-null.
   * If you don't, `outputArtifact` will be `null`.
   *
   * @default the Action will not have an output artifact
   */
  readonly outputArtifactName?: string;
}

/**
 * CodePipeline test Action that uses AWS CodeBuild.
 */
export class CodeBuildTestAction extends codepipeline.TestAction {
  private readonly props: CodeBuildTestActionProps;

  constructor(props: CodeBuildTestActionProps) {
    super({
      ...props,
      provider: 'CodeBuild',
      artifactBounds: { minInputs: 1, maxInputs: 5, minOutputs: 0, maxOutputs: 5 },
      configuration: {
        ProjectName: props.project.projectName,
      },
    });

    this.props = props;

    handleAdditionalInputOutputArtifacts(props, this,
      // pass functions to get around protected members
      this.actionInputArtifacts,
      (artifact) => this.addInputArtifact(artifact),
      (artifactName) => this.addOutputArtifact(artifactName));
  }

  /**
   * Returns the additional output artifacts defined for this Action.
   * Their names will be taken from the {@link CommonCodeBuildActionProps#additionalOutputArtifactNames}
   * property.
   *
   * @returns all additional output artifacts defined for this Action
   * @see #additionalOutputArtifact
   */
  public additionalOutputArtifacts(): codepipeline.Artifact[] {
    return this.outputArtifact === undefined
      ? this.actionOutputArtifacts
      : this.actionOutputArtifacts.slice(1);
  }

  /**
   * Returns the additional output artifact with the given name,
   * or throws an exception if an artifact with that name was not found
   * in the additonal output artifacts.
   * The names are defined by the {@link CommonCodeBuildActionProps#additionalOutputArtifactNames}
   * property.
   *
   * @param name the name of the artifact to find
   * @returns the artifact with the given name
   * @see #additionalOutputArtifacts
   */
  public additionalOutputArtifact(name: string): codepipeline.Artifact {
    return findOutputArtifact(this.additionalOutputArtifacts(), name);
  }

  protected bind(info: codepipeline.ActionBind): void {
    setCodeBuildNeededPermissions(info, this.props.project, this.actionOutputArtifacts.length > 0);
  }
}

function setCodeBuildNeededPermissions(info: codepipeline.ActionBind, project: codebuild.IProject,
                                       needsPipelineBucketWrite: boolean): void {
  // grant the Pipeline role the required permissions to this Project
  info.role.addToPolicy(new iam.PolicyStatement()
    .addResource(project.projectArn)
    .addActions(
      'codebuild:BatchGetBuilds',
      'codebuild:StartBuild',
      'codebuild:StopBuild',
    ));

  // allow the Project access to the Pipline's artifact Bucket
  if (needsPipelineBucketWrite) {
    info.pipeline.grantBucketReadWrite(project);
  } else {
    info.pipeline.grantBucketRead(project);
  }
}

function handleAdditionalInputOutputArtifacts(props: CommonCodeBuildActionProps, action: codepipeline.Action,
                                              inputArtifacts: codepipeline.Artifact[],
                                              addInputArtifact: (_: codepipeline.Artifact) => void,
                                              addOutputArtifact: (_: string) => void) {
  if ((props.additionalInputArtifacts || []).length > 0) {
    // we have to set the primary source in the configuration
    action.configuration.PrimarySource = inputArtifacts[0].artifactName;
    // add the additional artifacts
    for (const additionalInputArtifact of props.additionalInputArtifacts || []) {
      addInputArtifact(additionalInputArtifact);
    }
  }

  for (const additionalArtifactName of props.additionalOutputArtifactNames || []) {
    addOutputArtifact(additionalArtifactName);
  }
}

function findOutputArtifact(artifacts: codepipeline.Artifact[], name: string): codepipeline.Artifact {
  const ret = artifacts.find((artifact) => artifact.artifactName === name);
  if (!ret) {
    throw new Error(`Could not find output artifact with name '${name}'`);
  }
  return ret;
}
