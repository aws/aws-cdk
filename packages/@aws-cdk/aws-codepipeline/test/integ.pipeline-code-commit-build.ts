import codebuild = require('@aws-cdk/aws-codebuild');
import codecommit = require('@aws-cdk/aws-codecommit');
import cdk = require('@aws-cdk/cdk');
import codepipeline = require('../lib');

const app = new cdk.App();

const stack = new cdk.Stack(app, 'aws-cdk-codepipeline-codecommit-codebuild');

const repository = new codecommit.Repository(stack, 'MyRepo', {
  repositoryName: 'my-repo',
});
const sourceAction = new codecommit.PipelineSourceAction({
  actionName: 'source',
  outputArtifactName: 'SourceArtifact',
  repository,
  pollForSourceChanges: true,
});

const project = new codebuild.Project(stack, 'MyBuildProject', {
  source: new codebuild.CodePipelineSource(),
});
const buildAction = new codebuild.PipelineBuildAction({
  actionName: 'build',
  project,
  inputArtifact: sourceAction.outputArtifact,
});
const testAction = new codebuild.PipelineTestAction({
  actionName: 'test',
  project,
  inputArtifact: sourceAction.outputArtifact,
});

new codepipeline.Pipeline(stack, 'Pipeline', {
  stages: [
    {
      name: 'source',
      actions: [sourceAction],
    },
  ],
}).addStage({
  name: 'build',
  actions: [
    buildAction,
    testAction,
  ],
});

app.run();
