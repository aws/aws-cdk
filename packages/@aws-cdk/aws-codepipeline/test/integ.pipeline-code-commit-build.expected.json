{
  "Resources": {
    "MyRepoF4F48043": {
      "Type": "AWS::CodeCommit::Repository",
      "Properties": {
        "RepositoryName": "my-repo",
        "Triggers": []
      }
    },
    "PipelineArtifactsBucket22248F97": {
      "Type": "AWS::S3::Bucket",
      "DeletionPolicy": "Retain"
    },
    "PipelineRoleD68726F7": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "codepipeline.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        }
      }
    },
    "PipelineRoleDefaultPolicyC7A05455": {
      "Type": "AWS::IAM::Policy",
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject*",
                "s3:Abort*"
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "PipelineArtifactsBucket22248F97",
                    "Arn"
                  ]
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "PipelineArtifactsBucket22248F97",
                          "Arn"
                        ]
                      },
                      "/*"
                    ]
                  ]
                }
              ]
            },
            {
              "Action": [
                "codecommit:GetBranch",
                "codecommit:GetCommit",
                "codecommit:UploadArchive",
                "codecommit:GetUploadArchiveStatus",
                "codecommit:CancelUploadArchive"
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "MyRepoF4F48043",
                  "Arn"
                ]
              }
            },
            {
              "Action": [
                "codebuild:BatchGetBuilds",
                "codebuild:StartBuild",
                "codebuild:StopBuild"
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "MyBuildProject30DB9D6E",
                  "Arn"
                ]
              }
            },
            {
              "Action": [
                "codebuild:BatchGetBuilds",
                "codebuild:StartBuild",
                "codebuild:StopBuild"
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "MyBuildProject30DB9D6E",
                  "Arn"
                ]
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "PolicyName": "PipelineRoleDefaultPolicyC7A05455",
        "Roles": [
          {
            "Ref": "PipelineRoleD68726F7"
          }
        ]
      }
    },
    "PipelineC660917D": {
      "Type": "AWS::CodePipeline::Pipeline",
      "Properties": {
        "ArtifactStore": {
          "Location": {
            "Ref": "PipelineArtifactsBucket22248F97"
          },
          "Type": "S3"
        },
        "RoleArn": {
          "Fn::GetAtt": [
            "PipelineRoleD68726F7",
            "Arn"
          ]
        },
        "Stages": [
          {
            "Actions": [
              {
                "ActionTypeId": {
                  "Category": "Source",
                  "Owner": "AWS",
                  "Provider": "CodeCommit",
                  "Version": "1"
                },
                "Configuration": {
                  "RepositoryName": {
                    "Fn::GetAtt": [
                      "MyRepoF4F48043",
                      "Name"
                    ]
                  },
                  "BranchName": "master",
                  "PollForSourceChanges": true
                },
                "InputArtifacts": [],
                "Name": "source",
                "OutputArtifacts": [
                  {
                    "Name": "SourceArtifact"
                  }
                ],
                "RunOrder": 1
              }
            ],
            "Name": "source"
          },
          {
            "Actions": [
              {
                "ActionTypeId": {
                  "Category": "Build",
                  "Owner": "AWS",
                  "Provider": "CodeBuild",
                  "Version": "1"
                },
                "Configuration": {
                  "ProjectName": {
                    "Ref": "MyBuildProject30DB9D6E"
                  }
                },
                "InputArtifacts": [
                  {
                    "Name": "SourceArtifact"
                  }
                ],
                "Name": "build",
                "OutputArtifacts": [
                  {
                    "Name": "Artifact_build_awscdkcodepipelinecodecommitcodebuildMyBuildProject20018D4F"
                  }
                ],
                "RunOrder": 1
              },
              {
                "ActionTypeId": {
                  "Category": "Test",
                  "Owner": "AWS",
                  "Provider": "CodeBuild",
                  "Version": "1"
                },
                "Configuration": {
                  "ProjectName": {
                    "Ref": "MyBuildProject30DB9D6E"
                  }
                },
                "InputArtifacts": [
                  {
                    "Name": "SourceArtifact"
                  }
                ],
                "Name": "test",
                "OutputArtifacts": [],
                "RunOrder": 1
              }
            ],
            "Name": "build"
          }
        ]
      },
      "DependsOn": [
        "PipelineRoleDefaultPolicyC7A05455",
        "PipelineRoleD68726F7"
      ]
    },
    "MyBuildProjectRole6B7E2258": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "codebuild.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        }
      }
    },
    "MyBuildProjectRoleDefaultPolicy5604AA87": {
      "Type": "AWS::IAM::Policy",
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents"
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition"
                      },
                      ":logs:",
                      {
                        "Ref": "AWS::Region"
                      },
                      ":",
                      {
                        "Ref": "AWS::AccountId"
                      },
                      ":log-group:/aws/codebuild/",
                      {
                        "Ref": "MyBuildProject30DB9D6E"
                      }
                    ]
                  ]
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition"
                      },
                      ":logs:",
                      {
                        "Ref": "AWS::Region"
                      },
                      ":",
                      {
                        "Ref": "AWS::AccountId"
                      },
                      ":log-group:/aws/codebuild/",
                      {
                        "Ref": "MyBuildProject30DB9D6E"
                      },
                      ":*"
                    ]
                  ]
                }
              ]
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject*",
                "s3:Abort*"
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "PipelineArtifactsBucket22248F97",
                    "Arn"
                  ]
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "PipelineArtifactsBucket22248F97",
                          "Arn"
                        ]
                      },
                      "/*"
                    ]
                  ]
                }
              ]
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*"
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "PipelineArtifactsBucket22248F97",
                    "Arn"
                  ]
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "PipelineArtifactsBucket22248F97",
                          "Arn"
                        ]
                      },
                      "/*"
                    ]
                  ]
                }
              ]
            }
          ],
          "Version": "2012-10-17"
        },
        "PolicyName": "MyBuildProjectRoleDefaultPolicy5604AA87",
        "Roles": [
          {
            "Ref": "MyBuildProjectRole6B7E2258"
          }
        ]
      }
    },
    "MyBuildProject30DB9D6E": {
      "Type": "AWS::CodeBuild::Project",
      "Properties": {
        "Artifacts": {
          "Type": "CODEPIPELINE"
        },
        "Environment": {
          "ComputeType": "BUILD_GENERAL1_SMALL",
          "Image": "aws/codebuild/ubuntu-base:14.04",
          "PrivilegedMode": false,
          "Type": "LINUX_CONTAINER"
        },
        "ServiceRole": {
          "Fn::GetAtt": [
            "MyBuildProjectRole6B7E2258",
            "Arn"
          ]
        },
        "Source": {
          "Type": "CODEPIPELINE"
        }
      }
    }
  }
}
